### PR TITLE
Refine disable Fma

### PR DESCRIPTION
**Pull request by AMD-dwang**
Move the option “disableFma” from PipelineOptions to  PipelineShaderOptions.

### CODE REVIEW

File: include/vkgcDefs.h [hunk 1]
```diff
@@ -82,7 +82,7 @@
 //  %Version History
 //  | %Version | Change Description                                                                                    |
 //  | -------- | ----------------------------------------------------------------------------------------------------- |
-//  |     61.7 | Add disableFMA to PipelineOptions                                                                     |
+//  |     61.7 | Add disableFma to PipelineShaderOptions                                                               |
```
Comment (tsymalla) [inline include/vkgcDefs.h:85]: disableFma => disableFMA

File: include/vkgcDefs.h [hunk 2]
```diff
@@ -563,7 +563,6 @@ struct PipelineOptions {
 #endif
   unsigned forceNonUniformResourceIndexStageMask; ///< Mask of the stage to force using non-uniform resource index.
   bool reserved16;
-  bool disableFMA; ///< Disable to use FMA intrinsic and use FMUL + FADD instead.
```
Comment (tsymalla) [inline include/vkgcDefs.h:566]: Disable the use of FMA intrinsic

File: lgc/interface/lgc/Pipeline.h [hunk 3]
```diff
@@ -156,7 +156,6 @@ struct Options {
 #endif
   bool enableUberFetchShader; // Enable UberShader
   bool reserved16;
-  bool disableFMA; // Disable to use FMA intrinsic and use FMUL + FADD instead.
```
Comment (tsymalla) [inline lgc/interface/lgc/Pipeline.h:159]: Disable the use of FMA intrinsic

File: tool/dumper/vkgcPipelineDumper.cpp [hunk 4]
```diff
@@ -647,6 +647,7 @@ void PipelineDumper::dumpPipelineShaderInfo(const PipelineShaderInfo *shaderInfo
   dumpFile << "options.aggressiveInvariantLoads = " << shaderInfo->options.aggressiveInvariantLoads << "\n";
   dumpFile << "options.workaroundStorageImageFormats = " << shaderInfo->options.workaroundStorageImageFormats << "\n";
   dumpFile << "options.workaroundInitializeOutputsToZero = " << shaderInfo->options.workaroundInitializeOutputsToZero << "\n";
+  dumpFile << "options.disableFma = " << shaderInfo->options.disableFma << "\n";
```
Comment (tsymalla) [inline tool/dumper/vkgcPipelineDumper.cpp:650]: disableFMA

**Review by tsymalla — COMMENTED** [review]
LGTM

**Comment by tsymalla** [comment]
shaderdb/general/PipelineVsFs_DisableFMA.pipe is currently failing

**Comment by amdvlk-admin** [comment]
d5b1912deff8141b21784ebe7bdacd22dc9cea76 Jenkins build error.
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/tool/dumper/vkgcPipelineDumper.cpp:650:62: error: ‘const struct Vkgc::PipelineShaderOptions’ has no member named ‘disableFma’; did you mean ‘disableFMA’?
   dumpFile <<  options.disableFma =   << shaderInfo->options.disableFma <<   n ;
                                                              ^~~~~~~~~~
                                                              disableFMA
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/tool/dumper/vkgcPipelineDumper.cpp: In static member function ‘static void Vkgc::PipelineDumper::updateHashForPipelineShaderInfo(Vkgc::ShaderStage, const Vkgc::PipelineShaderInfo*, bool, Vkgc::PipelineDumper::MetroHash64*, bool)’:
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/tool/dumper/vkgcPipelineDumper.cpp:1675:30: error: ‘const struct Vkgc::PipelineShaderOptions’ has no member named ‘disableFma’; did you mean ‘disableFMA’?
       hasher->Update(options.disableFma);
                              ^~~~~~~~~~
                              disableFMA
[1109/3054] Building RISCVTargetParserDef.inc...
[1110/3054] Building CXX object compiler/llpc/llvm/third-party/benchmark/src/CMakeFiles/benchmark.dir/statistics.cc.o

**Comment by AMD-dwang** [comment]
> shaderdb/general/PipelineVsFs_DisableFMA.pipe is currently failing

Thanks. Fixed the failure.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 164849dd6c1af382ff3d5727cd296d481ef311f8
<details><summary>CTS tests (Failed: 0/215998)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40685/72006 (56.5%)</li>
<li>Failed: 0/72006 (0.0%)</li>
<li>Not Supported: 31321/72006 (43.5%)</li><li>Warnings: 0/72006 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40713/71984 (56.6%)</li>
<li>Failed: 0/71984 (0.0%)</li>
<li>Not Supported: 31271/71984 (43.4%)</li><li>Warnings: 0/71984 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Passed: 36765/72008 (51.1%)</li>
<li>Failed: 0/72008 (0.0%)</li>
<li>Not Supported: 35243/72008 (48.9%)</li><li>Warnings: 0/72008 (0.0%)</li>
</ul></details>


</ul></details>

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 5]
```diff
@@ -8741,6 +8744,12 @@ Value *SPIRVToLLVM::transGLSLExtInst(SPIRVExtInst *extInst, BasicBlock *bb) {
 
   case GLSLstd450Fma:
     // Fused multiply and add
+    if (m_shaderOptions->disableFMA || DisableFMA) {
+      auto fmf = getBuilder()->getFastMathFlags();
+      fmf.setAllowContract(false);
+      getBuilder()->setFastMathFlags(fmf);
+      return getBuilder()->CreateFAdd(getBuilder()->CreateFMul(args[0], args[1]), args[2]);
+    }
```
Comment (nhaehnle) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:8752]: I believe there should be an `IRBuilderBase::FastMathFlagGuard` here.

**Review by nhaehnle — COMMENTED** [review]
Thanks!

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp (hunk 5 above)
Comment (AMD-dwang) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp]: Fixed. Thanks.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 042d1e04137ae9e1367bb0ab9bda48b5f9660c74
<details><summary>CTS tests (Failed: 0/215993)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40695/72001 (56.5%)</li>
<li>Failed: 0/72001 (0.0%)</li>
<li>Not Supported: 31306/72001 (43.5%)</li><li>Warnings: 0/72001 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40713/71984 (56.6%)</li>
<li>Failed: 0/71984 (0.0%)</li>
<li>Not Supported: 31271/71984 (43.4%)</li><li>Warnings: 0/71984 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Passed: 36765/72008 (51.1%)</li>
<li>Failed: 0/72008 (0.0%)</li>
<li>Not Supported: 35243/72008 (48.9%)</li><li>Warnings: 0/72008 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — APPROVED** [review]

**Pull request merged**